packages/core/src/routes/session/middleware/passwordless-action.ts

```diff
@@ -0,0 +1,149 @@
+import { PasscodeType } from '@logto/schemas';
+import { MiddlewareType } from 'koa';
+import { Provider } from 'oidc-provider';
+
+import RequestError from '@/errors/RequestError';
+import { assignInteractionResults } from '@/lib/session';
+import { generateUserId, insertUser } from '@/lib/user';
+import { WithLogContext } from '@/middleware/koa-log';
+import {
+  hasUserWithPhone,
+  hasUserWithEmail,
+  findUserByPhone,
+  findUserByEmail,
+  updateUserById,
+} from '@/queries/user';
+import assertThat from '@/utils/assert-that';
+
+import { smsSessionResultGuard, emailSessionResultGuard } from '../types';
+import {
+  getVerificationStorageFromInteraction,
+  getPasswordlessRelatedLogType,
+  checkValidateExpiration,
+} from '../utils';
+
+export const smsSignInAction = <StateT, ContextT extends WithLogContext, ResponseBodyT>(
+  provider: Provider
+): MiddlewareType<StateT, ContextT, ResponseBodyT> => {
+  return async (ctx, next) => {
+    const verificationStorage = await getVerificationStorageFromInteraction(
+      ctx,
+      provider,
+      smsSessionResultGuard
+    );
+
+    const type = getPasswordlessRelatedLogType(PasscodeType.SignIn, 'sms');
+    ctx.log(type, verificationStorage);
+
+    const { phone, expiresAt } = verificationStorage;
+
+    checkValidateExpiration(expiresAt);
```

Comment:
do we need to check validate type? the verification result can only be used to specific action.